File: scripts/check-kit-interfaces.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2026 the Folio Project
# SPDX-License-Identifier: MIT

"""Check the published Kit SDK, then compile/link without repository header paths."""
import argparse
import plistlib
import re
import shutil
import subprocess
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
KITS = {"FolioKit": "FolioKit", "WriteKit": "Write",
        "ResearchKit": "Research", "ComposerKit": "Composer"}


def require(condition, message):
    if not condition:
        raise SystemExit(message)


def run(command, **kwargs):
    return subprocess.run(command, check=True, **kwargs)


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("products", type=Path, help="Xcode Build/Products/Debug or Release")
    args = parser.parse_args()
    products = args.products.resolve()
    public = {}
    owners = {}
    minimum_versions = set()
    for kit, project in KITS.items():
        module_map = (ROOT / project / kit / f"{kit}.modulemap").read_text()
        public[kit] = set(re.findall(r'^\s*header "([^"]+)"', module_map, re.M))
        require(public[kit], f"{kit}: missing explicit public header list")
        framework = products / f"{kit}.framework"
        with (framework / "Resources/Info.plist").open("rb") as stream:
            minimum_versions.add(plistlib.load(stream)["LSMinimumSystemVersion"])
        actual = {str(p.relative_to(framework / "Headers"))
                  for p in (framework / "Headers").rglob("*") if p.is_file()}
        require(actual == public[kit],
                f"{kit}: published headers differ: extra={actual - public[kit]}, missing={public[kit] - actual}")
        require((framework / "Modules/module.modulemap").read_text() == module_map,
                f"{kit}: build did not use the explicit module map")
        require(not any((framework / "PrivateHeaders").rglob("*.h")),
                f"{kit}: private headers must not be distributed to hosts")
        umbrella = (framework / "Headers" / f"{kit}.h").read_text()
        imported = set(re.findall(r'#import <' + kit + r'/([^>]+)>', umbrella))
        require(imported == public[kit] - {f"{kit}.h"},
                f"{kit}: umbrella and module map disagree")
        for header in (ROOT / project / kit).rglob("*.h"):
            owners[header.resolve()] = kit

    require(len(minimum_versions) == 1, "Kits must share a coordinated deployment target")
    minimum_version = minimum_versions.pop()

    # Owning applications get no special source-header access. A Kit's own
    # implementation may use local headers; all other callers use published paths.
    by_name = {path.name: (path, kit) for path, kit in owners.items()}
    for project in KITS.values():
        for source in (ROOT / project).rglob("*"):
            if source.suffix not in {".h", ".m", ".mm"}:
                continue
            own_kit = next((kit for kit, directory in KITS.items()
                            if source.is_relative_to(ROOT / directory / kit)), None)
            for token in re.findall(r'^\s*#\s*(?:import|include)\s*[<"]([^>"\n]+)[>"]', source.read_text(), re.M):
                known = by_name.get(Path(token).name)
                if known is None:
                    continue
                header, owner = known
                if own_kit == owner:
                    continue
                require(token == f"{owner}/{header.name}" and header.name in public[owner],
                        f"{source.relative_to(ROOT)}: use {owner}'s public interface, not {token}")

    require(not (products / "WriteKit.framework/Versions/Current/Frameworks/libFWEditor.dylib").exists(),
            "FWEditor must be compiled into WriteKit, not embedded")
    with tempfile.TemporaryDirectory(prefix="folio-kit-interface-") as temporary:
        stage = Path(temporary)
        # Only published bundles are visible; no Xcode header maps, source trees,
        # standalone implementation dylibs or private library search paths.
        for kit in KITS:
            shutil.copytree(products / f"{kit}.framework", stage / f"{kit}.framework", symlinks=True)
        sdk = subprocess.check_output(["xcrun", "--sdk", "macosx", "--show-sdk-path"], text=True).strip()
        common = ["xcrun", "clang", "-isysroot", sdk, "-fobjc-arc", "-fmodules",
                  f"-mmacosx-version-min={minimum_version}",
                  f"-fmodules-cache-path={stage / 'ModuleCache'}", "-F", str(stage),
                  "-Werror", "-Werror=non-modular-include-in-framework-module"]
        frameworks = [flag for kit in KITS for flag in ("-framework", kit)]
        # Check conventional umbrella imports and Clang module imports separately.
        fixture = (ROOT / "scripts/interface-checks/KitConsumer.m").read_text()
        for modules in (False, True):
            code = fixture
            if modules:
                for kit in KITS:
                    code = code.replace(f"#import <{kit}/{kit}.h>", f"@import {kit};")
            source = stage / "Consumer.m"
            source.write_text(code)
            run(common + [str(source), "-framework", "AppKit"] + frameworks +
                ["-Wl,-fatal_warnings", "-o", str(stage / "consumer")], cwd=stage)
        for header, kit in owners.items():
            if header.name in public[kit]:
                continue
            source = stage / "PrivateImport.m"
            source.write_text(f"#import <{kit}/{header.name}>\n")
            result = subprocess.run(common + ["-fsyntax-only", str(source)], cwd=stage,
                                    text=True, capture_output=True)
            require(result.returncode != 0 and f"'{kit}/{header.name}' file not found" in result.stderr,
                    f"{kit}: private header import was not rejected as expected: {header.name}\n{result.stderr}")
        for name in ("FKModelFoundations", "FKPackageSupport", "FKXMLSupport", "FWManuscript", "FWEditor"):
            source = stage / "PrivateModule.m"
            source.write_text(f"@import {name};\n")
            result = subprocess.run(common + ["-fsyntax-only", str(source)], cwd=stage,
                                    text=True, capture_output=True)
            require(result.returncode != 0 and f"module '{name}' not found" in result.stderr,
                    f"Private library module must not be importable: {name}\n{result.stderr}")
    print("Kit interfaces passed: exact published headers, caller imports, isolated compile/link, and private import rejection.")


if __name__ == "__main__":
    main()

File: scripts/check-kit-interfaces.rb
#!/usr/bin/env ruby
# SPDX-FileCopyrightText: 2026 the Folio Project
# SPDX-License-Identifier: MIT

require 'fileutils'
require 'json'
require 'open3'
require 'optparse'
require 'set'
require 'tmpdir'

# Check the published Kit SDK, then compile/link without repository header paths.
class KitInterfaceCheck
  ROOT = File.expand_path('..', __dir__)
  KITS = { 'FolioKit' => 'FolioKit', 'WriteKit' => 'Write',
           'ResearchKit' => 'Research', 'ComposerKit' => 'Composer' }.freeze
  IMPLEMENTATIONS = { 'FolioKit' => %w[FKModelFoundations FKPackageSupport FKXMLSupport],
                      'WriteKit' => %w[FWManuscript FWEditor] }.freeze

  def check(condition, message)
    raise message unless condition
  end

  def capture(*command)
    output, error, status = Open3.capture3(*command)
    check(status.success?, "#{command.first} failed: #{error}")
    output
  end

  def run(arguments)
    parser = OptionParser.new do |options|
      options.banner = 'Usage: ruby scripts/check-kit-interfaces.rb PRODUCTS'
      options.separator 'Check published Kit interfaces in Xcode Build/Products/Debug or Release.'
    end
    parser.parse!(arguments)
    check(arguments.length == 1, parser.to_s)
    products = File.expand_path(arguments.first)
    public_headers = {}
    owners = {}
    minimum_versions = Set.new
    KITS.each do |kit, project|
      module_map = File.read("#{ROOT}/#{project}/#{kit}/#{kit}.modulemap")
      public_headers[kit] = module_map.scan(/^\s*header "([^"]+)"/).flatten.to_set
      check(!public_headers[kit].empty?, "#{kit}: missing explicit public header list")
      framework = "#{products}/#{kit}.framework"
      info = JSON.parse(capture('plutil', '-convert', 'json', '-o', '-', "#{framework}/Resources/Info.plist"))
      minimum_versions.add(info.fetch('LSMinimumSystemVersion'))
      actual = Dir.glob("#{framework}/Headers/**/*", File::FNM_DOTMATCH)
                  .select { |path| File.file?(path) }.map { |path| path.delete_prefix("#{framework}/Headers/") }.to_set
      check(actual == public_headers[kit],
            "#{kit}: published headers differ: extra=#{(actual - public_headers[kit]).to_a.sort}, missing=#{(public_headers[kit] - actual).to_a.sort}")
      check(File.read("#{framework}/Modules/module.modulemap") == module_map,
            "#{kit}: build did not use the explicit module map")
      check(Dir.glob("#{framework}/PrivateHeaders/**/*.h").empty?, "#{kit}: private headers must not be distributed to hosts")
      imported = File.read("#{framework}/Headers/#{kit}.h").scan(/#import <#{kit}\/([^>]+)>/).flatten.to_set
      check(imported == public_headers[kit] - ["#{kit}.h"], "#{kit}: umbrella and module map disagree")
      Dir.glob("#{ROOT}/#{project}/#{kit}/**/*.h").each { |path| owners[File.realpath(path)] = kit }
    end
    check(minimum_versions.size == 1, 'Kits must share a coordinated deployment target')

    # A Kit's implementation may use local headers; all other callers use published paths.
    by_name = owners.each_with_object({}) { |(path, kit), result| result[File.basename(path)] = [path, kit] }
    KITS.each_value do |project|
      Dir.glob("#{ROOT}/#{project}/**/*.{h,m,mm}").each do |source|
        own_kit = KITS.find { |kit, directory| source.start_with?("#{ROOT}/#{directory}/#{kit}/") }&.first
        File.read(source).scan(/^\s*#\s*(?:import|include)\s*[<"]([^>"\n]+)[>"]/).flatten.each do |token|
          known = by_name[File.basename(token)]
          next unless known
          header, owner = known
          next if own_kit == owner
          name = File.basename(header)
          check(token == "#{owner}/#{name}" && public_headers[owner].include?(name),
                "#{source.delete_prefix(ROOT + '/')}: use #{owner}'s public interface, not #{token}")
        end
      end
    end
    IMPLEMENTATIONS.each do |kit, names|
      names.each do |name|
        check(Dir.glob("#{products}/#{kit}.framework/**/lib#{name}.dylib").empty?,
              "#{name} must compile into #{kit}, not remain embedded")
      end
    end
    Dir.mktmpdir('folio-kit-interface-') do |stage|
      # Only published bundles are visible, without source trees or Xcode header maps.
      KITS.each_key { |kit| FileUtils.cp_r("#{products}/#{kit}.framework", stage, preserve: true) }
      sdk = capture('xcrun', '--sdk', 'macosx', '--show-sdk-path').strip
      common = ['xcrun', 'clang', '-isysroot', sdk, '-fobjc-arc', '-fmodules',
                "-mmacosx-version-min=#{minimum_versions.first}", "-fmodules-cache-path=#{stage}/ModuleCache",
                '-F', stage, '-Werror', '-Werror=non-modular-include-in-framework-module']
      frameworks = KITS.keys.flat_map { |kit| ['-framework', kit] }
      fixture = File.read("#{ROOT}/scripts/interface-checks/KitConsumer.m")
      [false, true].each do |modules|
        code = fixture.dup
        KITS.each_key { |kit| code.gsub!("#import <#{kit}/#{kit}.h>", "@import #{kit};") } if modules
        source = "#{stage}/Consumer.m"
        File.write(source, code)
        check(system(*common, source, '-framework', 'AppKit', *frameworks,
                     '-Wl,-fatal_warnings', '-o', "#{stage}/consumer", chdir: stage), 'Kit consumer compile/link failed')
      end
      owners.each do |header, kit|
        name = File.basename(header)
        next if public_headers[kit].include?(name)
        source = "#{stage}/PrivateImport.m"
        File.write(source, "#import <#{kit}/#{name}>\n")
        _, error, status = Open3.capture3(*common, '-fsyntax-only', source, chdir: stage)
        check(!status.success? && error.include?("'#{kit}/#{name}' file not found"),
              "#{kit}: private header import was not rejected as expected: #{name}\n#{error}")
      end
      IMPLEMENTATIONS.values.flatten.each do |name|
        source = "#{stage}/PrivateModule.m"
        File.write(source, "@import #{name};\n")
        _, error, status = Open3.capture3(*common, '-fsyntax-only', source, chdir: stage)
        check(!status.success? && error.include?("module '#{name}' not found"),
              "Implementation folder must not be a public module: #{name}\n#{error}")
      end
    end
    puts 'Kit interfaces passed: exact published headers, caller imports, isolated compile/link, and private import rejection.'
  end
end

if $PROGRAM_NAME == __FILE__
  begin
    KitInterfaceCheck.new.run(ARGV)
  rescue StandardError => error
    warn error.message
    exit 1
  end
end
